### PR TITLE
Firewallns

### PR DESCRIPTION
**Pull request by NaveenSidhartha**
Added 'chains' to combine 'priority', 'policy' and 'rules' together as they are components of chains in nftables.

### CODE REVIEW

File: cloudcix_primitives/firewallns.py [hunk 1]
```diff
@@ -93,14 +93,98 @@ def snat_rule(nat):
...
+    if valid_nats is False:
+        validated = False
+        messages_list.append(f'{messages[msg_index]} {";".join(errors)}')
+    return validated, messages_list
+
+
+def validate_sets(sets, messages, msg_index):
+    messages_list = []
+    set_names = []
+    valid_sets = True
```
Comment (maria-walsh) [inline cloudcix_primitives/firewallns.py:157]: Why are two booleans required? 

From what I am reading both will either be True or both will be False. I am not seeing a scenario where one is True and the other is False

File: cloudcix_primitives/firewallns.py [hunk 2]
```diff
@@ -93,14 +93,98 @@ def snat_rule(nat):
...
+        validated = False
+        messages_list.append(f'{messages[msg_index + 1]} {";".join(errors)}')
+
+    return validated, messages_list
+
+
+def validate_nats(nats, messages, msg_index):
+    messages_list = []
+    errors = []
+    valid_nats = True
```
Comment (maria-walsh) [inline cloudcix_primitives/firewallns.py:139]: Same question, two booleans which will always have the same value

File: README.md [hunk 3]
```diff
@@ -27,7 +27,21 @@ supported verbs:
...
+            - str
+          protocol: str
+          port: array
+            - str
+          action: bool
+          log: bool
+          iiface: str
+          oiface: str
+          order: int
+    - user_rules: optional array
```
[the remark this answers is not in the document]
Comment (NaveenSidhartha) [reply README.md]: No, will discuss in the Scrum

File: cloudcix_primitives/templates/firewallns/nftables.conf.j2 [hunk 4]
```diff
@@ -67,57 +77,77 @@ table inet {{ table }} {
...
+    }
+{% endif %}
+{# ---------------------NAT Post Routing--------------------------------- #}
+{% if nat_postrouting is not None and nat_postrouting['lines']|length > 0 %}
+    # Outbound NAT(SNAT), Only IPv4 at Post-Routing stage,
+    # for nat type chains default policy is accept
+    chain nat_postrouting {
+        type nat hook postrouting priority {{ nat_postrouting['priority'] }}; policy accept;
+        # Source NAT 1:N
+    {% for line in nat_postrouting['lines'] %}
```
Comment (maria-walsh) [inline cloudcix_primitives/templates/firewallns/nftables.conf.j2:99]: https://github.com/CloudCIX/primitives/blob/main/cloudcix_primitives/templates/net/configs/netplan.yaml.j2

During the development of this template, we discovered that the jinja for and if statements should not be indented as it skews the formatting of what is rendered. 

Can you apply these lessons learned here?

File: cloudcix_primitives/firewallns.py [hunk 5]
```diff
@@ -345,44 +470,49 @@ def build(
...
-        3059: f'3059: Failed to connect to the Enabled PodNet for payload apply_nftables_file',
-        3060: f'3060: Failed to apply nftables file {nftables_file} on the Enabled PodNet',
-        3061: f'3061: Failed to connect to the Enabled PodNet for payload remove_nftables_file',
-        3062: f'3062: Failed to remove nftables file {nftables_file} on the Enabled PodNet',
+        3051: f'Failed to connect to the Enabled PodNet for payload create_nftables_file',
+        3052: f'Failed to create nftables file {nftables_file} on the Enabled PodNet',
+        3053: f'Failed to connect to the Enabled PodNet for payload validate_nftables_file',
+        3054: f'Failed to validate nftables file {nftables_file} on the Enabled PodNet',
+        3055: f'Failed to connect to the Enabled PodNet for payload read_table',
+        3056: f'Failed to read table {table} on the Enabled PodNet',
```
Comment (maria-walsh) [inline cloudcix_primitives/firewallns.py:496]: The message code is not used below.

File: cloudcix_primitives/firewallns.py [hunk 6]
```diff
@@ -345,44 +470,49 @@ def build(
...
-        3079: f'3079: Failed to connect to the Disabled PodNet for payload apply_nftables_file',
-        3080: f'3080: Failed to apply nftables file {nftables_file} on the Disabled PodNet',
-        3081: f'3081: Failed to connect to the Disabled PodNet for payload remove_nftables_file',
-        3082: f'3082: Failed to apply nftables file {nftables_file} on the Disabled PodNet',
+        3071: f'Failed to connect to the Disabled PodNet for payload create_nftables_file',
+        3072: f'Failed to create nftables file {nftables_file} on the Disabled PodNet',
+        3073: f'Failed to connect to the Disabled PodNet for payload validate_nftables_file',
+        3074: f'Failed to validate nftables file {nftables_file} on the Disabled PodNet',
+        3075: f'Failed to connect to the Disabled PodNet for payload read_table',
+        3076: f'Failed to read table {table} on the Disabled PodNet',
```
Comment (maria-walsh) [inline cloudcix_primitives/firewallns.py:509]: message code not used

File: cloudcix_primitives/firewallns.py [hunk 7]
```diff
@@ -648,16 +845,16 @@ def scrub(
...
-        3123: f'3123: Failed to connect to the Enabled PodNet for payload flush_table',
-        3124: f'3124: Failed to flush table {table} on the Enabled PodNet',
-        3131: f'3131: Failed to connect to the Disabled PodNet for payload read_table',
-        3132: f'3132: Failed to read table {table} on the Disabled PodNet',
-        3133: f'3133: Failed to connect to the Disabled PodNet for payload flush_table',
-        3134: f'3134: Failed to flush table {table} on the Disabled PodNet',
+        1100: f'Successfully removed nftables {table} in namespace {namespace}',
+        3100: f'Failed to remove nftables table {table} in namespace {namespace}',
+        3121: f'Failed to connect to the Enabled PodNet for payload read_table',
+        3122: f'Failed to read table {table} on the Enabled PodNet',
```
Comment (maria-walsh) [inline cloudcix_primitives/firewallns.py:851]: message code not used

File: cloudcix_primitives/firewallns.py [hunk 8]
```diff
@@ -648,16 +845,16 @@ def scrub(
...
+        3123: f'Failed to connect to the Enabled PodNet for payload flush_table',
+        3124: f'Failed to flush table {table} on the Enabled PodNet',
+        3131: f'Failed to connect to the Disabled PodNet for payload read_table',
+        3132: f'Failed to read table {table} on the Disabled PodNet',
```
Comment (maria-walsh) [inline cloudcix_primitives/firewallns.py:855]: message code not used

File: cloudcix_primitives/firewallns.py [hunk 9]
```diff
@@ -767,12 +964,12 @@ def read(
...
     # Define message
     messages = {
-        1200: f'1200: Successfully read nftables {table} in namespace {namespace}',
-        3200: f'3200: Failed to read nftables {table} in namespace {namespace}',
-        3221: f'3221: Failed to connect to the Enabled PodNet for payload read_table',
-        3222: f'3222: Failed to read table {table} from the Enabled PodNet',
-        3231: f'3231: Failed to connect to the Disabled PodNet for payload read_table',
-        3232: f'3232: Failed to read table {table} from the Disabled PodNet',
+        1200: f'Successfully read nftables {table} in namespace {namespace}',
+        3200: f'Failed to read nftables {table} in namespace {namespace}',
```
Comment (maria-walsh) [inline cloudcix_primitives/firewallns.py:968]: message code not used

**Review by maria-walsh — APPROVED** [review]

**Pull request merged**